pryvvid/aws-api-gateway-lambda
# Task9

1. [Task 9](https://github.com/EPAM-JS-Competency-center/cloud-development-course-initial/blob/main/task9-bff-elastic-beanstalk/task.md)
2. PR 02.11.2021 / Deadline 06.11.2021
3. Self-score: 7/7

## IMPORTANT
Elastic beanstalk enviroments will be terminated after cross-check (because I don't want to pay addtional money)

## Bff endpoints

Product-service bff:
http://pryvvid-bff-api-dev.eu-west-1.elasticbeanstalk.com/products

Cart-service bff:
http://pryvvid-bff-api-dev.eu-west-1.elasticbeanstalk.com/cart

## Api gateway endpoints:

Product-service:
https://fmfiacgetl.execute-api.eu-west-1.amazonaws.com/products

Cart-service bff:
https://nn9tlrdcij.execute-api.eu-west-1.amazonaws.com/dev/api/profile/cart

## Create product body example
```
    {
        "title": "New game test",
        "description": "My test game",
        "price": 20,
        "count": 10
    }
```

## Front-end react app
https://d38kvhmy9feixw.cloudfront.net/

## Evaluation criteria
3 - A working and correct express application should be in the bff-service folder. Reviewer can start this application locally with any valid configuration in the .env file and this application should works as described in the task 9.1
5 - The bff-service should be deployed with Elastic Beanstalk. The bff-service call should be redirected to the appropriate service : product-service or CART. The response from the bff-service should be the same as if product-service or CART services were called directly.
Additional (optional) tasks
+1 - Add a cache at the bff-service level for a request to the getProductsList function of the product-service. The cache should expire in 2 minutes.
How to test:
Get products list
Create new product
Get products list - result shouldn’t have new product
Wait more than 2 minutes
Get products list - result should have new product

**shouldReturnCachedData** util runs timer:
https://github.com/pryvvid/aws-api-gateway-lambda/blob/task9/bff-service/nestjs/src/utils/shouldReturnCachedData.ts

If the timer < 2 minutes, the data comes from cache
https://github.com/pryvvid/aws-api-gateway-lambda/blob/1b7d28e4db7bcb17d90ebcdc3815b0d094da48a4/bff-service/nestjs/src/app.service.ts#L18
```js
    if (req.originalUrl.substring(1) === 'products' && req.method === 'GET') {
      if (shouldReturnCachedData())
        return res.status(this.cachedStatus).json(this.cachedProducts);
    }
```

+1 - Use NestJS to create bff-service instead of express
https://github.com/pryvvid/aws-api-gateway-lambda/tree/task9/bff-service/nestjs
